---
title: "Combine counts results"
author: "Jia Nie"
date: "11/3/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

install packages
```{r package, include=FALSE}
packageList <- c("rio", "dplyr", "tibble")
for (package in packageList) {
  if(!require(package, character.only = TRUE)){
    install.packages(package); require(package, character.only = TRUE); }
}
```

# Combine Metrics Summary to Counts Records
```{r chunk 1}
Sequencing <- rio::import("/media/jianie/DATA/UTHSC_cellranger_codes_files/ExperimentSummaryFiles/Form 2_Sequencing Samples.xlsx")
aggr <- rio::import("/media/jianie/DATA/UTHSC_cellranger_codes_files/ExperimentSummaryFiles/Form 3_cellranger_counts_summary.xlsx")
aggr2 <- aggr[-which(aggr$Status %in% c("Delete", "middle")), ]
df.all <- data.frame()
for (pp in aggr2$Metrics) {
  dx <- rio::import(paste0(pp))
  dx[] <- lapply(dx, function(x){ if(is.integer(x)) as.numeric(x) else x} )
  dx[] <- lapply(dx, function(x){ if(is.double(x)) as.numeric(as.integer(x)) else x} )
  dx[] <- lapply(dx, function(x){ if(is.factor(x)) as.numeric(as.character(x)) else x} )
  dx[] <- lapply(dx[], function(x){ if(is.character(x)) as.numeric(gsub("%|,", "", x)) else x} )
  # sapply(dx, class)
  # sapply(dx, mode)
  dx <- add_column(dx, sample_id_ori = aggr2[which(aggr2$Metrics == pp), "sample_id_ori"], .before = 1)
  dx <- add_column(dx, Counts_List = aggr2[which(aggr2$Metrics == pp), "Counts_List"], .before = 1)
  if ( match(pp, aggr2$Metrics) == 1 ) {df.all <- rbind(df.all, dx) } else {
    if ( length( union(setdiff(colnames(df.all), colnames(dx)), setdiff(colnames(dx), colnames(df.all))) ) > 0 ) {
      df.all <- dplyr::bind_rows(df.all, dx)} else {
      df.all <- rbind(df.all, dx)} } 
  # print(pp)
}

intersect(colnames(df.all), colnames(aggr2))
df.all2 <- merge(df.all, aggr2, by = c("Counts_List", "sample_id_ori"))
intersect(colnames(df.all2), colnames(Sequencing))
df.all3 <- merge(df.all2, Sequencing, by = c("sample_id", "Sequencing_Date", "Project", "Lane", "SQ_List", "Disk", "CountsFolder"))
colnames(df.all3)

# add sum for sequencing lane
sum <- df.all3 %>% filter(Status == "Original") %>% group_by(Sequencing_Date, Lane) %>% summarise(Sum.Sequencing = sum(`Number of Reads`))
df.all3 <- merge(sum, df.all3, by = c("Sequencing_Date", "Lane"))
df.all3 <- add_column(df.all3, Percent.PerLane = ifelse(df.all3$Status == "Original", 
                                                        round(df.all3$`Number of Reads`/df.all3$Sum.Sequencing *100, digits = 1), NA), 
                      .after = 3)
df.all3 <- add_column(df.all3, Diff.PerLane = ifelse(df.all3$Status == "Original", 
                                                        round(df.all3$Percent.PerLane - (df.all3$`Target %` *100), digits = 1), NA), 
                      .after = 4)
df.all3 <- df.all3[order(df.all3$SQ_List),]
colnames(df.all3)
columns.select <- c("Counts_List", "sample_id_ori", "SQ_List" ,  "Sequencing_Date", "Lane" , 
                    "sample_id",  "Project" ,"Disk" , "CountsFolder" , "sample_id" ,
                    "Sequencing_Date", "Project" ,"Lane", "SQ_List", "Disk",
                    "CountsFolder" ,"Counts_List" , "sample_id_ori" , "Estimated Number of Cells" ,
                    "Mean Reads per Cell" ,"Median Genes per Cell" , "Number of Reads", "Cell Numbers",  "Reads/Cell", "Reads/Sample",
                    "Target %" , "Sum" ,"Agg" , "Status", "molecule_h5",  
                    "Sample_Date",   "Library_ID",  "I7_Index ID", "I7_Index_Sequence" ,
                    "I5_Index ID" , "I5_Index_Sequence")
df.depth <- df.all3 %>% select(all_of(columns.select))
```


```{r chunk 2}
saveRDS(df.all3, file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/ExperimentSummaryFiles/Form 4_Counts_Sequencing_Combine.Rds")
write.csv(df.all3, file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/ExperimentSummaryFiles/Form 4_Counts_Sequencing_Combine.csv", row.names = FALSE)
write.csv(df.depth, file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/ExperimentSummaryFiles/Form 4_Counts_Sequencing_Combine_simple.csv", row.names = FALSE)
```




# Select rows from project for Aggregation in Cellranger
```{r}
cols2 = c("sample_id", "molecule_h5", "sample_id_ori2", "Counts_List")
df.all3 <- readRDS("/media/jianie/DATA/UTHSC_cellranger_codes_files/ExperimentSummaryFiles/Form 4_Counts_Sequencing_Combine.Rds")
unique(df.all3$Sequencing_Date)
table(df.all3$Sequencing_Date, df.all3$Disk)
```

# Seagate_SequencingFiles_1 -----------------------------------------------------------------------------------------------------------
```{r}
S1 <- df.all3[which((df.all3$Sequencing_Date %in% c("20190613", "20190717", "20190816", "20190912", "20191121", "20210114"))), ]
S1 <- S1 %>% arrange(sample_id)  %>% rename(sample_id_ori2 = sample_id,
                                                sample_id = sample_id_ori)
write.csv(S1[cols2], file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/AggrFiles/S1.agg.csv", row.names=FALSE)
write.csv(S1, file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/AggrFiles/S1.csv", row.names=FALSE)
```


# Seagate_SequencingFiles_2 -----------------------------------------------------------------------------------------------------------
```{r}
S2 <- df.all3[which((df.all3$Sequencing_Date %in% c("20210830", "20210716", "20210126"))), ]
S2 <- S2 %>% arrange(sample_id)  %>% rename(sample_id_ori2 = sample_id,
                                                sample_id = sample_id_ori)
write.csv(S2[cols2], file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/AggrFiles/S2.agg.csv", row.names=FALSE)
write.csv(S2, file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/AggrFiles/S2.csv", row.names=FALSE)
```



# SequencingFiles_3 -------------------------------------------------------------------------------------------------------------------
## 20211105_L3_L4
```{r}
S3_1 <- df.all3[which((df.all3$Sequencing_Date %in% c("20211105"))), ]
S3_1 <- S3_1 %>% arrange(sample_id)  %>% rename(sample_id_ori2 = sample_id,
                                                sample_id = sample_id_ori)
write.csv(S3_1[cols2], file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/AggrFiles/S3_1.agg.csv", row.names=FALSE)
write.csv(S3_1, file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/AggrFiles/S3_1.csv", row.names=FALSE)
```

## 20220616_L1_Counts, 20220620_Counts
```{r}
S3_2 <- df.all3[which((df.all3$Sequencing_Date %in% c("20220616", "20220620"))), ]
S3_2 <- S3_2 %>% arrange(sample_id)  %>% rename(sample_id_ori2 = sample_id,
                                                sample_id = sample_id_ori)
write.csv(S3_2[cols2], file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/AggrFiles/S3_2.agg.csv", row.names=FALSE)
write.csv(S3_2, file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/AggrFiles/S3_2.csv", row.names=FALSE)
```


# SequencingFiles_4 -------------------------------------------------------------------------------------------------------------------
```{r}
S4 <- df.all3[which((df.all3$Sequencing_Date %in% c("20220720"))), ]
S4 <- S4 %>% arrange(sample_id)  %>% rename(sample_id_ori2 = sample_id,
                                                sample_id = sample_id_ori)
write.csv(S4[cols2], file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/AggrFiles/S4.agg.csv", row.names=FALSE)
write.csv(S4, file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/AggrFiles/S4.csv", row.names=FALSE)
```









# Project based
```{r}
cols = c("sample_id", "molecule_h5", "sample_id_ori", "Counts_List")
df.all3 <- readRDS("/media/jianie/DATA/UTHSC_cellranger_codes_files/ExperimentSummaryFiles/Form 4_Counts_Sequencing_Combine.Rds")
```

STARR all project
```{r}
STARR.All <- df.all3[which(df.all3$Project == "STARR" & df.all3$Agg %in% c("Yes", "Tobe")), ]
STARR.All <- STARR.All %>% arrange(sample_id)
write.csv(STARR.All[cols], file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/AggrFiles/STARR.All.agg.csv", row.names=FALSE)
```


STARR project
```{r}
STARR <- df.all3[which(df.all3$Project == "STARR" & df.all3$Agg == "Yes"), ]
STARR <- STARR %>% arrange(sample_id)
write.csv(STARR[cols], file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/AggrFiles/STARR.agg.csv", row.names=FALSE)

```

STARR project_FZ_LS_test
```{r}
cols2 = c("sample_id", "molecule_h5", "sample_id_ori2", "Counts_List")
Test <- df.all3[which(df.all3$Project == "STARR" & df.all3$sample_id_ori %in% c("STARR_077", 
                                                                                "STARR_077_LS", "STARR_077_LS_force12248",
                                                                                "STARR_077_FZ", "STARR_077_FZ_force12248")), ]
Test <- Test %>% arrange(sample_id)  %>% rename(sample_id_ori2 = sample_id,
                                                sample_id = sample_id_ori)
write.csv(Test[cols2], file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/AggrFiles/STARR_FZ_Test.agg.csv", row.names=FALSE)
```


SGLT2 project
```{r}
SGLT2 <- df.all3[which(df.all3$Project == "SGLT2" & df.all3$Agg == "Yes"), ]
SGLT2 <- SGLT2 %>% arrange(sample_id)
write.csv(SGLT2[cols], file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/AggrFiles/SGLT2.agg.csv", row.names=FALSE)
```

STARR & SGLT2-V5 project
```{r}
STARR2 <- df.all3[which((df.all3$Project == "STARR" | df.all3$sample_id %in% c("76615_V5", "76632_V5", "76638_V5", "76643_V5",
                                                                              "76647_V5", "76649_V5", "76656_V5", "76658_V5",
                                                                              "76660_V5", "76661_V5", "76662_V5", "76664_V5")) & 
                         df.all3$Agg == "Yes") , ]
STARR2 <- STARR2 %>% arrange(sample_id)
write.csv(STARR2[cols], file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/AggrFiles/STARR2.agg.csv", row.names=FALSE)
```

STARR & SGLT2 project
```{r}
STARR2 <- df.all3[which(df.all3$Project %in% c("SGLT2", "STARR") & df.all3$Agg == "Yes") , ]
STARR2 <- STARR2 %>% arrange(sample_id)
write.csv(STARR2[cols], file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/AggrFiles/STARR2.agg.csv", row.names=FALSE)
```


Masternak project
```{r}
dataframe <- df.all3[which(df.all3$Project == "Masternak" ), ] # & df.all3$Agg == "Yes"
write.csv(dataframe[cols], file = "/media/jianie/DATA/UTHSC_cellranger_codes_files/AggrFiles/Masternak.agg.csv", row.names=FALSE)
```



